[RPC] [Minor] RPC entry point cleanup

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34292 [RPC] [Minor] RPC entry point cleanup**

This is to finish a cleanup request from https://github.com/pytorch/pytorch/pull/34733#discussion_r392479110.

Differential Revision: [D7436759](https://our.internmc.facebook.com/intern/diff/D7436759/)

xush6528 committed:
[RPC] Increase ref count before releasing GIL

I noticed, after https://github.com/pytorch/pytorch/pull/33610, there is incorrect ref count on the objects passed as RPC args, specially, those being wrapped into `py::args`.

I think `const py::args&` and `const py::kwargs&` reference to a `py::object` held by other Python thread, so if we release GIL in the C++ thread, other Python work could continue, thus the ref count is messed up.

Differential Revision: [D7436759](https://our.internmc.facebook.com/intern/diff/D7436759/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -448,8 +448,8 @@ If the future completes with an error, an exception is thrown.
          const std::string& qualifiedNameStr,
-         const py::args& args,
-         const py::kwargs& kwargs) {
+         py::args args,
```
Is it guaranteed behavior that pybind won't release GIL when passing args?

xush6528 replied:
Yes, I think pybind holds GIL when preparing `py::args`, and it also hold GIL when destructing `py::args` (which de-refs the contained `py::object`s).

Since if we write,

```cpp
m.def("foo", foo, py::call_guard<gil_scoped_release>());
```

It's equivalent to this, (according to pybind doc, https://pybind11.readthedocs.io/en/stable/reference.html#_CPPv410call_guard),
where GIL is released only within the local scope of user function.
```
m.def("foo", [](args...) {
    gil_scopede_release scope_guard;
    return foo(args...); // Notice after returning, `scope_guard` is destructed before `args` being desturcted, so GIL is re-acquired before destructing `args`.
});
```

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -448,8 +448,8 @@ If the future completes with an error, an exception is thrown.
-         const py::args& args,
-         const py::kwargs& kwargs) {
+         py::args args,
+         py::kwargs kwargs) {
```
Shall add a NOLINT to get rid of lint failure and add some comments to explain?

xush6528 replied:
Yes. Updated.

xush6528 committed:
Update on "[WIP] [RPC] Increase ref count before releasing GIL"


I noticed, after https://github.com/pytorch/pytorch/pull/33610, there is incorrect ref count on the objects passed as RPC args, specially, those being wrapped into `py::args`.

I think `const py::args&` and `const py::kwargs&` reference to a `py::object` held by other Python thread, so if we release GIL in the C++ thread, other Python work could continue, thus the ref count is messed up.

Differential Revision: [D7436759](https://our.internmc.facebook.com/intern/diff/D7436759/)

[ghstack-poisoned]

xush6528 committed:
Update on "[WIP] [RPC] Increase ref count before releasing GIL"


I noticed, after https://github.com/pytorch/pytorch/pull/33610, there is incorrect ref count on the objects passed as RPC args, specially, those being wrapped into `py::args`.

I think `const py::args&` and `const py::kwargs&` reference to a `py::object` held by other Python thread, so if we release GIL in the C++ thread, other Python work could continue, thus the ref count is messed up.

Differential Revision: [D7436759](https://our.internmc.facebook.com/intern/diff/D7436759/)

[ghstack-poisoned]

mrshenli reviewed (approved):
Thanks this LGTM! Let's add some comment to explain the reason for creating copies of py::args and using NOLINT. Thanks for fixing!!

xush6528 committed:
Update on "[Debug] py::object ref count issue"


I noticed, after https://github.com/pytorch/pytorch/pull/33610, there is incorrect ref count on the objects passed as RPC args, specially, those being wrapped into `py::args`.

I think `const py::args&` and `const py::kwargs&` reference to a `py::object` held by other Python thread, so if we release GIL in the C++ thread, other Python work could continue, thus the ref count is messed up.

Differential Revision: [D7436759](https://our.internmc.facebook.com/intern/diff/D7436759/)

[ghstack-poisoned]

xush6528 commented:
This change used to cause hanging in tests, test_return_local_script_class_rref_in_py_and_use_in_script, test_return_local_script_module_rref_in_py_and_use_in_script, which should be resulted form ref count update without GIL.

With fixes in the stack of https://github.com/pytorch/pytorch/pull/38364, the hanging can't be reproduced anymore. Merging it now.

xush6528 committed:
Update on "[RPC] [Minor] RPC entry point cleanup"


This is to finish a cleanup request from https://github.com/pytorch/pytorch/pull/34733#discussion_r392479110.

Differential Revision: [D7436759](https://our.internmc.facebook.com/intern/diff/D7436759/)

[ghstack-poisoned]
